### PR TITLE
cli: alter use subcommands for patch

### PR DESCRIPTION
**Pull request by slack-coder**
Make `rad patch` accept subcommands for specifying its operation in place of the '--command' notation.

For instance `rad patch update` versus `rad patch --update`.  Also make the default operation list active patches, replacing patch creation. Patch creation is now done via `rad patch open`.

Make these changes in a way consistent with other commands like `rad delegate`.

Add '--no-confirm' to patch to support this subcommand's documentation test.

References: #148

### CODE REVIEW

File: radicle-cli/src/commands/patch.rs [hunk 1]
```diff
@@ -24,79 +24,93 @@ pub const HELP: Help = Help {
...
 "#,
 };
 
-#[derive(Debug)]
-pub enum Update {
-    No,
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum OptPatch {
     Any,
+    None,
```
Comment (cloudhead) [inline radicle-cli/src/commands/patch.rs:48]: Doesn't really work because then you can get an update with `None` as the patch_id, which is nonsensical. I'd remove this type completely and just have `Update { patch_id: Option<PatchId> }`

File: radicle-cli/src/commands/patch.rs [hunk 2]
```diff
@@ -129,17 +143,43 @@ impl Args for Options {
                 Long("help") => {
                     return Err(Error::Help.into());
                 }
+
+                Value(val) if op.is_none() => match val.to_string_lossy().as_ref() {
+                    "list" => op = Some(OperationName::List),
+                    "open" => op = Some(OperationName::Create),
+                    "update" => op = Some(OperationName::Update),
```
Comment (cloudhead) [inline radicle-cli/src/commands/patch.rs:150]: Let's also support short options "l" "o" "u"

File: radicle-cli/src/commands/patch.rs [hunk 3]
```diff
@@ -24,79 +24,93 @@ pub const HELP: Help = Help {
...
     fn default() -> Self {
-        Self::No
+        Self::List
     }
 }
 
-#[derive(Default, Debug)]
+#[derive(Debug)]
+pub enum Operation {
+    Create {
```
Comment (cloudhead) [inline radicle-cli/src/commands/patch.rs:73]: Rename to `Open` along with `OperationName`

File: radicle-cli/src/commands/patch/create.rs [hunk 4]
```diff
@@ -194,14 +205,14 @@ pub fn run(
     term::patch::list_commits(&commits)?;
     term::blank();
 
-    if !term::confirm("Continue?") {
+    if !confirm("Continue?", &options) {
         anyhow::bail!("patch proposal aborted by user");
     }
 
-    let message = head_commit
+    let commit_message = head_commit
```
Comment (cloudhead) [inline radicle-cli/src/commands/patch/create.rs:212]: Why was this renamed?

File: radicle-cli/src/commands/patch.rs [hunk 5]
```diff
@@ -129,17 +143,43 @@ impl Args for Options {
...
+                    "list" => op = Some(OperationName::List),
+                    "open" => op = Some(OperationName::Create),
+                    "update" => op = Some(OperationName::Update),
+
+                    unknown => anyhow::bail!("unknown operation '{}'", unknown),
+                },
+                Value(val) if op == Some(OperationName::Update) && id == OptPatch::Any => {
+                    let val = val
+                        .to_str()
+                        .ok_or_else(|| anyhow!("issue id specified is not UTF-8"))?;
```
Comment (cloudhead) [inline radicle-cli/src/commands/patch.rs:157]:
```suggestion
                        .ok_or_else(|| anyhow!("patch id specified is not UTF-8"))?;
```

**Comment by cloudhead** [comment]
Looks good overall :+1:

File: radicle-cli/src/commands/patch.rs (hunk 3 above)
Comment (slack-coder) [reply radicle-cli/src/commands/patch.rs:73]: I was following `rad issue`s lead on this one.  We should change there as well if we do here.

https://github.com/radicle-dev/heartwood/blob/master/radicle-cli/src/commands/issue.rs#L147

File: radicle-cli/src/commands/patch.rs (hunk 1 above)
Comment (slack-coder) [reply radicle-cli/src/commands/patch.rs:48]: Its messy because working on the code required some surgery.  This is an update on the original code which does the same thing.

Both `update` and `create` use a single method whose operations relies on this argument.  Refactoring this meaningfully appeared outside of the issue's scope.

File: radicle-cli/src/commands/patch/create.rs [hunk 6]
```diff
@@ -33,11 +35,18 @@ blank is also okay.
 -->
 "#;
 
+#[inline]
+fn confirm<D: fmt::Display>(prompt: D, options: &Options) -> bool {
+    !options.confirm || term::confirm(prompt)
+}
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply radicle-cli/src/commands/patch/create.rs:41]: Yeah maybe you're right. Just feels weird to pass options into this function as well, but we can live with it. I think it might be worth having a prompter that takes a boolean and then knows whether to prompt or not, eg.
```rust
let prompt = Prompter::new(options.confirm);
if prompt.confirm("Are you sure?") {
...
}
```

Could be a nice re-usable component if we use this pattern a lot.

[the remark this answers is not in the document]
Comment (slack-coder) [reply radicle-cli/src/commands/patch/create.rs:48]: What do you think about something like this for now?

https://github.com/radicle-dev/heartwood/pull/214/commits/9d8e3e3bfddf178cf4db103a3dcf87ddd21717df

[the remark this answers is not in the document]
Comment (cloudhead) [reply radicle-cli/src/commands/patch/create.rs:48]: no no no, this is very arcane

[the remark this answers is not in the document]
Comment (cloudhead) [reply radicle-cli/src/commands/patch/create.rs:48]: I'd leave it how you have it for now

**Comment by slack-coder** [comment]
@cloudhead all points addressed except for one issue

File: radicle-cli/src/commands/patch/create.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cloudhead) [reply radicle-cli/src/commands/patch/create.rs:41]: Yeah, objects with methods are more common and idiomatic if you need to return something from a function

**Review by cloudhead — APPROVED** [review]
LGTM; what's the issue with the failing test?

**Comment by slack-coder** [comment]
> LGTM; what's the issue with the failing test?

Ive been trying to figure this out.  It fails to push the branch to rad. I cant figure how why theres a descrepancy, maybe the git version requires this on the integration test.

!! Pushing HEAD to storage...
== Command failed
fatal: The current branch flux-capacitor-power has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream rad flux-capacitor-power

**Comment by cloudhead** [comment]
And it works outside of the testing environment?

**Comment by slack-coder** [comment]
@cloudhead I've made some changes, please confirm the last two commits.

I'll get cracking on a proper fix for the documentation test dependency on `radicle-remote-helper`

File: radicle-cli/src/commands/patch/create.rs [hunk 7]
```diff
@@ -66,9 +74,11 @@ pub fn run(
...
     let head_oid = head.target().ok_or(anyhow!("invalid HEAD ref; aborting"))?;
     let head_commit = workdir.find_commit(head_oid)?;
-    let head_branch = head
-        .shorthand()
-        .ok_or(anyhow!("cannot create patch from detached head; aborting"))?;
+    let head_branch = if head.is_branch() {
+        git::raw::Branch::wrap(head)
+    } else {
+        anyhow::bail!("cannot create patch from detached head; aborting")
+    };
```
[the remark this answers is not in the document]
Comment (slack-coder) [reply radicle-cli/src/commands/patch/create.rs:81]: Having the branch as a `Branch` instead of a `str` is useful for inspecting it (line 102) and is stricter/more expressive.

**Comment by slack-coder** [comment]
The `rad patch` issue this references has a '--closes' argument not supported in these changes.

It will be added in an upcoming PR.
